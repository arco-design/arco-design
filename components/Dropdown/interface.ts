import { CSSProperties, ReactNode } from 'react';
import { TriggerProps } from '../Trigger';
import { ButtonProps } from '../Button';

/**
 * @title Dropdown
 */
export interface DropdownProps {
  children?: ReactNode;
  /**
   * @zh 下拉框的内容
   * @en Content of dropdown list
   */
  droplist?: ReactNode;
  /**
   * @zh 下拉框的弹出位置
   * @en Position of dropdown list
   * @defaultValue bl
   */
  position?: 'top' | 'tl' | 'tr' | 'bottom' | 'bl' | 'br';
  /**
   * @zh 触发下拉框弹出的方式
   * @en Types of events that cause the popup to show
   * @defaultValue hover
   */
  trigger?: TriggerProps['trigger'];
  /**
   * @zh 是否禁用弹出
   * @en Whether to disable popup
   * @version 2.16.0
   */
  disabled?: boolean;
  /**
   * @zh 隐藏后是否销毁 DOM 结构
   * @en Whether to umount the node on hiding
   * @defaultValue true
   */
  unmountOnExit?: boolean;
  /**
   * @zh 控制下拉框是否默认打开
   * @en Whether the dropdown list is visible by default
   */
  defaultPopupVisible?: boolean;
  /**
   * @zh 控制下拉框是否打开（受控模式）
   * @en Whether the dropdown list is visible (Controlled)
   */
  popupVisible?: boolean;
  /**
   * @zh 弹出框下可接受所有 `Trigger` 组件的 `Props`
   * @en Pass all `Trigger` component properties
   */
  triggerProps?: Partial<TriggerProps>;
  /**
   * @zh 弹出框打开/关闭时会触发
   * @en Callback when visibility changes
   */
  onVisibleChange?: (visible: boolean) => void;
  /**
   * @zh 弹出框挂在的父级节点
   * @en To set the container of the dropdown menu
   */
  getPopupContainer?: (node: HTMLElement) => Element;
}

/**
 * @title Dropdown.Button
 */
export interface DropdownButtonProps extends Omit<DropdownProps, 'position'> {
  style?: CSSProperties;
  className?: string | string[];
  /**
   * @zh 下拉框的弹出位置
   * @en Position of the dropdown list
   * @defaultValue br
   */
  position?: 'top' | 'tl' | 'tr' | 'bottom' | 'bl' | 'br';
  /**
   * @zh 等同于 `Button` 的 size
   * @en Same as `size` of ButtonProps
   */
  size?: 'mini' | 'small' | 'default' | 'large';
  /**
   * @zh 等同于 `Button` 的 type
   * @en Same as `type` of ButtonProps
   * @defaultValue default
   */
  type?: 'default' | 'primary' | 'secondary' | 'dashed' | 'outline' | 'text';
  /**
   * @zh 接收 button 按钮的所有 Props，应用于左侧按钮
   * @en Pass Button's properties to the button on the left
   */
  buttonProps?: ButtonProps;
  /**
   * @zh 右侧显示内容，可以是 icon 或者任意 dom 元素
   * @en Custom icon on the right
   * @defaultValue <IconMore />
   */
  icon?: ReactNode;
  /**
   * @zh 自定义两个按钮的渲染
   * @en Custom buttons inside Dropdown.Button
   */
  buttonsRender?: (buttons: ReactNode[]) => ReactNode[];
  /**
   * @zh 左侧按钮的点击事件
   * @en Callback when button on the left is clicked
   */
  onClick?: (e: Event) => void;
  children?: ReactNode;
}
