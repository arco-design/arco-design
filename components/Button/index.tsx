import React, { useState, useContext, useEffect, createRef, ReactNode, forwardRef } from 'react';
import IconLoading from '../../icon/react-icon/IconLoading';
import Group from './group';
import cs from '../_util/classNames';
import { ConfigContext } from '../ConfigProvider';
import { ButtonProps } from './interface';
import useMergeProps from '../_util/hooks/useMergeProps';

const regexTwoCNChar = /^[\u4e00-\u9fa5]{2}$/;

function processChildren(children?: ReactNode) {
  const childrenList = [];
  let isPrevChildPure = false;
  React.Children.forEach(children, (child) => {
    const isCurrentChildPure = typeof child === 'string' || typeof child === 'number';
    if (isCurrentChildPure && isPrevChildPure) {
      const lastIndex = childrenList.length - 1;
      const lastChild = childrenList[lastIndex];
      childrenList[lastIndex] = `${lastChild}${child}`;
    } else {
      childrenList.push(child);
    }
    isPrevChildPure = isCurrentChildPure;
  });
  return React.Children.map(childrenList, (child) =>
    typeof child === 'string' ? <span>{child}</span> : child
  );
}

const defaultProps: ButtonProps = {
  htmlType: 'button',
  type: 'default',
  shape: 'square',
};

function Button(baseProps: ButtonProps, ref) {
  const {
    getPrefixCls,
    size: ctxSize,
    autoInsertSpaceInButton,
    componentConfig,
  } = useContext(ConfigContext);
  const props = useMergeProps<ButtonProps>(baseProps, defaultProps, componentConfig?.Button);
  const {
    style,
    className,
    children,
    htmlType,
    type,
    status,
    size,
    shape,
    href,
    anchorProps,
    disabled,
    loading,
    loadingFixedWidth,
    icon,
    iconOnly,
    onClick,
    long,
    ...rest
  } = props;

  const iconNode = loading ? <IconLoading /> : icon;

  const [isTwoCNChar, setIsTwoCNChar] = useState(false);
  const buttonRef = ref || createRef();

  useEffect(() => {
    if (autoInsertSpaceInButton && buttonRef && buttonRef.current) {
      const textContent = buttonRef.current.textContent;
      if (regexTwoCNChar.test(textContent)) {
        if (!isTwoCNChar) {
          setIsTwoCNChar(true);
        }
      } else if (isTwoCNChar) {
        setIsTwoCNChar(false);
      }
    }
  }, [buttonRef, autoInsertSpaceInButton]);

  const prefixCls = getPrefixCls('btn');
  const _type = type === 'default' ? 'secondary' : type;
  const classNames = cs(
    prefixCls,
    `${prefixCls}-${_type}`,
    `${prefixCls}-size-${size || ctxSize}`,
    `${prefixCls}-shape-${shape}`,
    {
      [`${prefixCls}-long`]: long,
      [`${prefixCls}-status-${status}`]: status,
      [`${prefixCls}-loading-fixed-width`]: loadingFixedWidth,
      [`${prefixCls}-loading`]: loading,
      [`${prefixCls}-link`]: href,
      [`${prefixCls}-icon-only`]: iconOnly || (!children && children !== 0 && iconNode),
      [`${prefixCls}-disabled`]: disabled,
      [`${prefixCls}-two-chinese-chars`]: isTwoCNChar,
    },
    className
  );

  const handleClick: React.MouseEventHandler<HTMLElement> = (event: any): void => {
    if (loading) {
      typeof event?.preventDefault === 'function' && event.preventDefault();
      return;
    }
    onClick && onClick(event);
  };

  const InnerContent = (
    <>
      {iconNode}
      {processChildren(children)}
    </>
  );

  if (href) {
    const _anchorProps = { ...anchorProps };
    if (disabled) {
      delete _anchorProps.href;
    } else {
      _anchorProps.href = href;
    }
    return (
      <a
        ref={buttonRef}
        {...rest}
        {..._anchorProps}
        style={style}
        className={classNames}
        onClick={handleClick}
      >
        {InnerContent}
      </a>
    );
  }

  return (
    <button
      ref={buttonRef}
      {...rest}
      style={style}
      className={classNames}
      type={htmlType}
      disabled={disabled}
      onClick={handleClick}
    >
      {InnerContent}
    </button>
  );
}

const ForwardRefButton = forwardRef<unknown, ButtonProps>(Button);

const ButtonComponent = ForwardRefButton as typeof ForwardRefButton & {
  __BYTE_BUTTON: boolean;
  Group: typeof Group;
};

ButtonComponent.__BYTE_BUTTON = true;

ButtonComponent.Group = Group;

ButtonComponent.displayName = 'Button';

export default ButtonComponent;

export { ButtonProps };
