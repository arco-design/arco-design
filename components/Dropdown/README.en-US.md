`````
Component / Navigation

# Dropdown

When there are too many commands, the alternative commands can be stored in the floating container that expands downward.
`````

%%Content%%

## API

### Dropdown

|Property|Description|Type|DefaultValue|Version|
|---|---|---|---|---|
|defaultPopupVisible|Whether the dropdown list is visible by default|boolean |`-`|-|
|disabled|Whether to disable popup|boolean |`-`|2.16.0|
|popupVisible|Whether the dropdown list is visible (Controlled)|boolean |`-`|-|
|unmountOnExit|Whether to umount the node on hiding|boolean |`true`|-|
|position|Position of dropdown list|'top' \| 'tl' \| 'tr' \| 'bottom' \| 'bl' \| 'br' |`bl`|-|
|trigger|Types of events that cause the popup to show|[TriggerProps](trigger#trigger)['trigger'] |`hover`|-|
|droplist|Content of dropdown list|ReactNode |`-`|-|
|triggerProps|Pass all `Trigger` component properties|Partial&lt;[TriggerProps](trigger#trigger)&gt; |`-`|-|
|getPopupContainer|To set the container of the dropdown menu|(node: HTMLElement) => Element |`-`|-|
|onVisibleChange|Callback when visibility changes|(visible: boolean) => void |`-`|-|

### Dropdown.Button

|Property|Description|Type|DefaultValue|Version|
|---|---|---|---|---|
|disabled|Whether the dropdown menu is disabled|boolean |`-`|2.6.0|
|unmountOnExit|Whether to umount the node on hiding|boolean |`true`|-|
|position|Position of the dropdown list|'top' \| 'tl' \| 'tr' \| 'bottom' \| 'bl' \| 'br' |`br`|-|
|size|Same as `size` of ButtonProps|'mini' \| 'small' \| 'default' \| 'large' |`-`|-|
|trigger|Types of events that cause the popup to show|[TriggerProps](trigger#trigger)['trigger'] |`hover`|-|
|type|Same as `type` of ButtonProps|'default' \| 'primary' \| 'secondary' \| 'dashed' \| 'outline' \| 'text' |`default`|-|
|droplist|Content of the dropdown list|ReactNode |`-`|-|
|icon|Custom icon on the right|ReactNode |`<IconMore />`|-|
|buttonProps|Pass Button's properties to the button on the left|ButtonProps |`-`|-|
|className|Additional css class|string \| string[] |`-`|-|
|style|Additional style|CSSProperties |`-`|-|
|buttonsRender|Custom buttons inside Dropdown.Button|(buttons: ReactNode[]) => ReactNode[] |`-`|-|
|onClick|Callback when button on the left is clicked|(e: Event) => void |`-`|-|
|onVisibleChange|Callback when visibility changes|(visible: boolean) => void |`-`|-|
