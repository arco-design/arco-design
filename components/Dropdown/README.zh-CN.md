`````
组件 / 导航

# 下拉菜单 Dropdown

页面上的命令过多时，可将备选命令收纳到向下展开的浮层容器中。
`````

%%Content%%

## API

### Dropdown

|参数名|描述|类型|默认值|版本|
|---|---|---|---|---|
|defaultPopupVisible|控制下拉框是否默认打开|boolean |`-`|-|
|disabled|是否禁用弹出|boolean |`-`|2.16.0|
|popupVisible|控制下拉框是否打开（受控模式）|boolean |`-`|-|
|unmountOnExit|隐藏后是否销毁 DOM 结构|boolean |`true`|-|
|position|下拉框的弹出位置|'top' \| 'tl' \| 'tr' \| 'bottom' \| 'bl' \| 'br' |`bl`|-|
|trigger|触发下拉框弹出的方式|[TriggerProps](trigger#trigger)['trigger'] |`hover`|-|
|droplist|下拉框的内容|ReactNode |`-`|-|
|triggerProps|弹出框下可接受所有 `Trigger` 组件的 `Props`|Partial&lt;[TriggerProps](trigger#trigger)&gt; |`-`|-|
|getPopupContainer|弹出框挂在的父级节点|(node: HTMLElement) => Element |`-`|-|
|onVisibleChange|弹出框打开/关闭时会触发|(visible: boolean) => void |`-`|-|

### Dropdown.Button

|参数名|描述|类型|默认值|版本|
|---|---|---|---|---|
|disabled|禁用。|boolean |`-`|2.6.0|
|unmountOnExit|隐藏后是否销毁 DOM 结构|boolean |`true`|-|
|position|下拉框的弹出位置|'top' \| 'tl' \| 'tr' \| 'bottom' \| 'bl' \| 'br' |`br`|-|
|size|等同于 `Button` 的 size|'mini' \| 'small' \| 'default' \| 'large' |`-`|-|
|trigger|触发下拉框弹出的方式|[TriggerProps](trigger#trigger)['trigger'] |`hover`|-|
|type|等同于 `Button` 的 type|'default' \| 'primary' \| 'secondary' \| 'dashed' \| 'outline' \| 'text' |`default`|-|
|droplist|下拉框的内容|ReactNode |`-`|-|
|icon|右侧显示内容，可以是 icon 或者任意 dom 元素|ReactNode |`<IconMore />`|-|
|buttonProps|接收 button 按钮的所有 Props，应用于左侧按钮|ButtonProps |`-`|-|
|className|节点类名|string \| string[] |`-`|-|
|style|节点样式|CSSProperties |`-`|-|
|buttonsRender|自定义两个按钮的渲染|(buttons: ReactNode[]) => ReactNode[] |`-`|-|
|onClick|左侧按钮的点击事件|(e: Event) => void |`-`|-|
|onVisibleChange|弹出框打开/关闭时会触发|(visible: boolean) => void |`-`|-|
