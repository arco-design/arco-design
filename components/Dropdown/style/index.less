@import '../../style/mixin.less';
@import './token.less';

@dropdown-prefix-cls: ~'@{prefix}-dropdown';

.@{dropdown-prefix-cls} {
  &-menu {
    position: relative;
    box-sizing: border-box;
    max-height: @dropdown-max-height;
    padding: @dropdown-padding-vertical 0;
    border: 1px solid @dropdown-color-border;
    border-radius: @dropdown-border-radius;
    background-color: @dropdown-color-bg;
    box-shadow: @dropdown-box-shadow;
    overflow: auto;

    &-hidden {
      display: none;
    }

    &-item,
    &-pop-header {
      position: relative;
      box-sizing: border-box;
      width: 100%;
      height: @dropdown-option-height;
      padding: 0 @dropdown-option-padding-horizontal;
      font-size: @dropdown-font-size;
      line-height: $height;
      text-align: left;
      cursor: pointer;
      // 避免 Trigger 的小箭头位于其上方
      z-index: 1;
      .text-ellipsis();

      // default
      color: @dropdown-option-color-text_default;
      background-color: @dropdown-option-color-bg_default;

      // selected
      &.@{dropdown-prefix-cls}-menu-selected {
        color: @dropdown-option-color-text_selected;
        background-color: @dropdown-option-color-bg_selected;
        font-weight: @dropdown-option-font-weight_selected;
        transition: all @transition-duration-1 @transition-timing-function-linear;
      }

      // hover
      &:hover {
        color: @dropdown-option-color-text_hover;
        background-color: @dropdown-option-color-bg_hover;
      }

      // focus by Tab
      &:focus-visible {
        box-shadow: 0 0 0 2px @color-primary-6 inset;
      }

      // hotkey active
      &.@{dropdown-prefix-cls}-menu-active {
        box-shadow: 0 0 0 1px @color-primary-6 inset;
      }

      // disabled
      &.@{dropdown-prefix-cls}-menu-disabled {
        color: @dropdown-option-color-text_disabled;
        background-color: @dropdown-option-color-bg_disabled;
        cursor: not-allowed;
      }

      // overwrite color of links
      a,
      a:hover,
      a:focus,
      a:active {
        color: inherit;
        cursor: inherit;
        text-decoration: none;
      }

      > a:only-child::before {
        content: '';
        position: absolute;
        left: 0;
        right: 0;
        top: 0;
        bottom: 0;
      }
    }

    &-pop-header {
      display: flex;
      align-items: center;
      justify-content: space-between;

      .@{dropdown-prefix-cls}-menu-icon-suffix {
        margin-left: @dropdown-margin-left-suffix-icon;
      }
    }

    &-group {
      &:first-child &-title {
        margin-top: @dropdown-group-title-margin-top - @dropdown-padding-vertical;
      }
    }

    &-group-title {
      margin-top: @dropdown-group-title-margin-top;
      box-sizing: border-box;
      width: 100%;
      padding: 0 @dropdown-group-title-padding-horizontal;
      line-height: @dropdown-group-title-height;
      font-size: @dropdown-group-title-font-size;
      color: @dropdown-group-title-color-text;
      .text-ellipsis();
    }
  }

  &-menu-dark {
    border-color: @dropdown-dark-color-border;
    background-color: @dropdown-dark-color-bg;

    .@{dropdown-prefix-cls}-menu-item,
    .@{dropdown-prefix-cls}-menu-pop-header {
      // default
      color: @dropdown-dark-option-color-text_default;
      background-color: @dropdown-dark-option-color-bg_default;

      // selected
      &.@{dropdown-prefix-cls}-menu-selected {
        color: @dropdown-dark-option-color-text_selected;
        background-color: @dropdown-dark-option-color-bg_selected;

        &:hover {
          color: @dropdown-dark-option-color-text_selected;
        }
      }

      // hover
      &:hover {
        color: @dropdown-dark-option-color-text_hover;
        background-color: @dropdown-dark-option-color-bg_hover;
      }

      // disabled
      &.@{dropdown-prefix-cls}-menu-disabled {
        color: @dropdown-dark-option-color-text_disabled;
        background-color: @dropdown-dark-option-color-bg_disabled;
      }
    }

    .@{dropdown-prefix-cls}-menu-group-title {
      color: @dropdown-dark-group-title-color-text;
    }
  }

  &-menu-pop-trigger {
    // 下拉框的 menu 弹出不需要小箭头
    .@{prefix}-trigger-arrow {
      display: none;
    }
  }

  &-menu + .@{prefix}-trigger-arrow {
    background-color: @dropdown-color-bg;
  }
}

.@{dropdown-prefix-cls}-menu-rtl {
  .@{dropdown-prefix-cls}-menu {
    &-item,
    &-pop-header {
      text-align: right;

      .@{dropdown-prefix-cls}-menu-icon-suffix {
        margin-left: 0;
        margin-right: @dropdown-margin-left-suffix-icon;
      }
    }
  }
}
