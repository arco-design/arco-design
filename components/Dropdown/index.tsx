import React, { ReactElement, useContext, useRef } from 'react';
import Trigger, { EventsByTriggerNeed } from '../Trigger';
import Button from './button';
import { ConfigContext } from '../ConfigProvider';
import cs from '../_util/classNames';
import useMergeValue from '../_util/hooks/useMergeValue';
import omit from '../_util/omit';
import pick from '../_util/pick';
import { DropdownProps } from './interface';
import useMergeProps from '../_util/hooks/useMergeProps';

const defaultProps: DropdownProps = {
  position: 'bl',
  trigger: 'hover',
  unmountOnExit: true,
};

function Dropdown(baseProps: DropdownProps, _) {
  const { getPrefixCls, componentConfig } = useContext(ConfigContext);
  const props = useMergeProps<DropdownProps>(baseProps, defaultProps, componentConfig?.Button);
  const {
    trigger,
    droplist,
    children,
    position,
    disabled,
    unmountOnExit,
    triggerProps,
    getPopupContainer,
    onVisibleChange,
    ...rest
  } = props;

  const prefixCls = getPrefixCls('dropdown');

  const triggerRef = useRef(null);
  const [popupVisible, setPopupVisible] = useMergeValue(false, {
    defaultValue: props.defaultPopupVisible,
    value: props.popupVisible,
  });

  const getPopupContent = () => {
    return React.Children.only(droplist || <span />) as React.ReactElement;
  };

  const changePopupVisible = (visible: boolean) => {
    setPopupVisible(visible);
    onVisibleChange && onVisibleChange(visible);
    triggerProps && triggerProps.onVisibleChange && triggerProps.onVisibleChange(visible);
  };

  const handleVisibleChange = (visible: boolean) => {
    if (visible !== popupVisible) {
      changePopupVisible(visible);
    }
  };

  const renderPopup = () => {
    const content = getPopupContent();
    return content && content.props.isMenu
      ? React.cloneElement(content as ReactElement, {
          prefixCls: `${prefixCls}-menu`,
          inDropdown: true,
          selectable: false,
          onClickMenuItem: (key, event) => {
            let returnValueOfOnClickMenuItem = null;

            // Trigger onClickMenuItem first
            const content = getPopupContent();
            if (content.props.onClickMenuItem) {
              returnValueOfOnClickMenuItem = content.props.onClickMenuItem(key, event);
            }

            // Set focus to avoid onblur
            const child = triggerRef.current && triggerRef.current.getRootElement();
            child && child.focus && child.focus();

            // Trigger onVisibleChange. Outer component can determine whether to change the state based on the current visibility value.
            if (returnValueOfOnClickMenuItem instanceof Promise) {
              returnValueOfOnClickMenuItem.finally(() => changePopupVisible(false));
            } else if (returnValueOfOnClickMenuItem !== false) {
              changePopupVisible(false);
            }
          },
        })
      : content;
  };

  return (
    <Trigger
      ref={(ref) => (triggerRef.current = ref)}
      classNames="slideDynamicOrigin"
      childrenPrefix={prefixCls}
      trigger={trigger}
      popup={renderPopup}
      mouseEnterDelay={400}
      mouseLeaveDelay={400}
      disabled={disabled}
      unmountOnExit={unmountOnExit}
      position={position}
      popupVisible={popupVisible}
      popupAlign={{
        left: 4,
        right: 4,
        top: 4,
        bottom: 4,
      }}
      getPopupContainer={getPopupContainer}
      alignPoint={trigger === 'contextMenu'}
      {...pick(rest, EventsByTriggerNeed)}
      {...omit(triggerProps, ['onVisibleChange'])}
      onVisibleChange={handleVisibleChange}
    >
      {React.isValidElement(children)
        ? React.cloneElement(children, {
            disabled,
            className: cs(
              {
                [`${prefixCls}-popup-visible`]: popupVisible,
              },
              children.props.className
            ),
          })
        : children}
    </Trigger>
  );
}

const ForwardRefDropdown: React.ForwardRefExoticComponent<
  DropdownProps & React.RefAttributes<unknown>
> = React.forwardRef<unknown, DropdownProps>(Dropdown);

const DropdownComponent = ForwardRefDropdown as typeof ForwardRefDropdown & {
  Button: typeof Button;
};

DropdownComponent.displayName = 'Dropdown';

DropdownComponent.Button = Button;

export default DropdownComponent;

export { DropdownProps };
