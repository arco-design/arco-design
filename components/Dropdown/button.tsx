import React, { ReactNode, ReactElement, forwardRef, useContext } from 'react';
import ArcoButton from '../Button';
import Dropdown from './index';
import IconMore from '../../icon/react-icon/IconMore';
import { DropdownButtonProps } from './interface';
import { ConfigContext } from '../ConfigProvider';
import useMergeProps from '../_util/hooks/useMergeProps';
import { pickDataAttributes } from '../_util/pick';

const defaultProps: DropdownButtonProps = {
  position: 'br',
  trigger: 'hover',
  type: 'default',
  icon: <IconMore />,
  unmountOnExit: true,
};

function Button(baseProps: DropdownButtonProps, ref) {
  const { componentConfig } = useContext(ConfigContext);
  const props = useMergeProps<DropdownButtonProps>(
    baseProps,
    defaultProps,
    componentConfig?.['Dropdown.Button']
  );
  const {
    className,
    style,
    children,
    disabled,
    droplist,
    position = 'br',
    trigger = 'hover',
    type = 'default',
    size,
    icon = <IconMore />,
    unmountOnExit = true,
    onClick,
    onVisibleChange,
    buttonProps,
    buttonsRender,
  } = props;

  let leftButton: ReactNode = (
    <ArcoButton disabled={disabled} type={type} size={size} onClick={onClick} {...buttonProps}>
      {children}
    </ArcoButton>
  );
  let rightButton: ReactNode = (
    <ArcoButton disabled={disabled} type={type} size={size} icon={icon} />
  );

  if (buttonsRender) {
    [leftButton, rightButton] = buttonsRender([leftButton, rightButton]);
  }

  const disableTrigger =
    disabled ||
    !rightButton ||
    ((rightButton as ReactElement).props && (rightButton as ReactElement).props.loading);

  return (
    <ArcoButton.Group className={className} style={style} ref={ref} {...pickDataAttributes(props)}>
      {leftButton}
      <Dropdown
        disabled={disabled}
        droplist={droplist}
        position={position}
        trigger={trigger}
        unmountOnExit={unmountOnExit}
        onVisibleChange={onVisibleChange}
        triggerProps={{
          disabled: disableTrigger,
        }}
      >
        {rightButton}
      </Dropdown>
    </ArcoButton.Group>
  );
}

const ButtonComponent = forwardRef(Button);

ButtonComponent.displayName = 'DropdownButton';

export default ButtonComponent;

export { DropdownButtonProps };
