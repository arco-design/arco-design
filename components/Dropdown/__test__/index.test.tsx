import React from 'react';
import { mount } from 'enzyme';
import mountTest from '../../../tests/mountTest';
import componentConfigTest from '../../../tests/componentConfigTest';
import Dropdown from '..';
import Button from '../../Button';
import Menu from '../../Menu';

mountTest(Dropdown);
componentConfigTest(Dropdown.Button, 'Dropdown.Button');

function mountDropdown(component: React.ReactElement) {
  return mount(component);
}

const mockFn = jest.fn();
const Droplist = (
  <Menu onClickMenuItem={mockFn}>
    <Menu.Item key="1">Menu Item 1</Menu.Item>
    <Menu.Item key="2">Menu Item 2</Menu.Item>
    <Menu.Item key="3">Menu Item 3</Menu.Item>
  </Menu>
);

describe('Dropdown', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  it("dropdown's children can be string", () => {
    const wrapper = mountDropdown(<Dropdown droplist={Droplist}>Hello</Dropdown>);
    expect(wrapper.find('span').text()).toBe('Hello');
  });

  it('dropdown open correctly', () => {
    const wrapper = mountDropdown(
      <Dropdown droplist={Droplist} trigger="click">
        <Button>Click Me</Button>
      </Dropdown>
    );
    wrapper.find(Button).simulate('click');
    jest.runAllTimers();
    const menu = wrapper.find('Menu');
    expect(menu.length).toBe(1);
    expect(menu.find('MenuItem').length).toBe(3);
  });

  it('click menu item', () => {
    const wrapper = mountDropdown(
      <Dropdown
        trigger="click"
        droplist={
          <Menu
            onClickMenuItem={(key) => {
              if (key === '2') {
                return false;
              }

              if (key === '3') {
                return new Promise((resolve) => {
                  setTimeout(() => resolve(null), 100);
                });
              }
            }}
          >
            <Menu.Item key="1">Menu Item 1</Menu.Item>
            <Menu.Item key="2">Menu Item 2</Menu.Item>
            <Menu.Item key="3">Menu Item 3</Menu.Item>
          </Menu>
        }
      >
        <Button>Click Me</Button>
      </Dropdown>
    );

    const simulateMenuItemClick = (index: number) => {
      wrapper.find('Menu').find('MenuItem').at(index).simulate('click');
    };

    const getIsTriggerVisible = () => {
      return wrapper.find('Trigger').state().popupVisible;
    };

    expect(getIsTriggerVisible()).toBe(false);
    wrapper.find(Button).simulate('click');
    expect(getIsTriggerVisible()).toBe(true);

    simulateMenuItemClick(0);
    expect(getIsTriggerVisible()).toBe(false);

    wrapper.find(Button).simulate('click');
    simulateMenuItemClick(1);
    expect(getIsTriggerVisible()).toBe(true);

    simulateMenuItemClick(2);
    expect(getIsTriggerVisible()).toBe(true);
  });

  it('Dropdown.Button mount correctly', () => {
    const dropList = (
      <Menu>
        <Menu.Item key="1">Save now</Menu.Item>
        <Menu.Item key="2">Save and Publish</Menu.Item>
      </Menu>
    );
    const wrapper = mountDropdown(
      <Dropdown.Button droplist={dropList}>
        <Button>Click Me</Button>
      </Dropdown.Button>
    );

    expect(wrapper.find('.arco-btn-group')).toHaveLength(1);
  });
});
