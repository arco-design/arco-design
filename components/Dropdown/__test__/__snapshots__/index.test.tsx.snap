// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigProvider componentConfig.Dropdown.Button default 1`] = `
<div
  class="arco-btn-group"
>
  <button
    class="arco-btn arco-btn-secondary arco-btn-size-default arco-btn-shape-square"
    type="button"
  />
  <button
    class="arco-btn arco-btn-secondary arco-btn-size-default arco-btn-shape-square arco-btn-icon-only"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="arco-icon arco-icon-more"
      fill="none"
      focusable="false"
      stroke="currentColor"
      stroke-width="4"
      viewBox="0 0 48 48"
    >
      <path
        d="M38 25v-2h2v2h-2ZM23 25v-2h2v2h-2ZM8 25v-2h2v2H8Z"
        fill="currentColor"
        stroke="none"
      />
      <path
        d="M38 25v-2h2v2h-2ZM23 25v-2h2v2h-2ZM8 25v-2h2v2H8Z"
      />
    </svg>
  </button>
</div>
`;

exports[`ConfigProvider componentConfig.Dropdown.Button set className globally 1`] = `
<div
  class="arco-btn-group global-default-class"
>
  <button
    class="arco-btn arco-btn-secondary arco-btn-size-default arco-btn-shape-square"
    type="button"
  />
  <button
    class="arco-btn arco-btn-secondary arco-btn-size-default arco-btn-shape-square arco-btn-icon-only"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="arco-icon arco-icon-more"
      fill="none"
      focusable="false"
      stroke="currentColor"
      stroke-width="4"
      viewBox="0 0 48 48"
    >
      <path
        d="M38 25v-2h2v2h-2ZM23 25v-2h2v2h-2ZM8 25v-2h2v2H8Z"
        fill="currentColor"
        stroke="none"
      />
      <path
        d="M38 25v-2h2v2h-2ZM23 25v-2h2v2h-2ZM8 25v-2h2v2H8Z"
      />
    </svg>
  </button>
</div>
`;

exports[`ConfigProvider componentConfig.Dropdown.Button set className globally and component set className itself. 1`] = `
<div
  class="arco-btn-group component-class"
>
  <button
    class="arco-btn arco-btn-secondary arco-btn-size-default arco-btn-shape-square"
    type="button"
  />
  <button
    class="arco-btn arco-btn-secondary arco-btn-size-default arco-btn-shape-square arco-btn-icon-only"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="arco-icon arco-icon-more"
      fill="none"
      focusable="false"
      stroke="currentColor"
      stroke-width="4"
      viewBox="0 0 48 48"
    >
      <path
        d="M38 25v-2h2v2h-2ZM23 25v-2h2v2h-2ZM8 25v-2h2v2H8Z"
        fill="currentColor"
        stroke="none"
      />
      <path
        d="M38 25v-2h2v2h-2ZM23 25v-2h2v2h-2ZM8 25v-2h2v2H8Z"
      />
    </svg>
  </button>
</div>
`;
